procedural/scheduling/src/main/kotlin/gov/nasa/ammos/aerie/procedural/scheduling/utils/DefaultEditablePlanDriver.kt

```diff
@@ -0,0 +1,315 @@
+      if (directive.startTime > duration()) {
+        throw Exception("New activity with id ${directive.id.id()} would start after the end of the plan")
+      }
+      if (directive.start is DirectiveStart.Absolute && directive.startTime.isNegative) {
+        throw Exception("New activity with id ${directive.id.id()} would start before the beginning of the plan")
+      }
+    }
+  }
+
+  private data class Commit(
+    val diff: Set<Edit>,
+
+    /**
+     * A record of the simulation results objects that were up-to-date when the commit
+     * was created.
+     *
+     * This has SHARED OWNERSHIP with [DefaultEditablePlanDriver]; the editable plan may add more to
+     * this list AFTER the commit is created.
+     */
+    val upToDateSimResultsSet: MutableSet<WeakReference<PerishableSimulationResults>>
+  )
+
+  private var committedChanges = Commit(setOf(), mutableSetOf())
+  private var uncommittedChanges = mutableListOf<Edit>()
+
+  /** Whether there are uncommitted changes. */
+  val isDirty
+    get() = uncommittedChanges.isNotEmpty()
+
+  /** The total reduced set of changes made to the plan. */
+  val totalDiff: Set<Edit>
+    get() = committedChanges.diff
+
+  // Jointly owned set of up-to-date simulation results. See class-level comment for algorithm explanation.
+  private var upToDateSimResultsSet: MutableSet<WeakReference<PerishableSimulationResults>> = mutableSetOf()
+
+  override fun latestResults(): SimulationResults? {
+    val internalResults = adapter.latestResults()
+
+    // kotlin checks structural equality by default, not referential equality.
+    val isStale = internalResults?.inputDirectives()?.toSet() != directives().toSet()
+
+    internalResults?.setStale(isStale)
+
+    if (!isStale) upToDateSimResultsSet.add(WeakReference(internalResults))
+    return internalResults
+  }
+
+  override fun create(directive: NewDirective): ActivityDirectiveId {
+    class ParentSearchException(id: ActivityDirectiveId, size: Int): Exception("Expected one parent activity with id $id, found $size")
+    val id = adapter.generateDirectiveId()
+    val parent = when (val s = directive.start) {
+      is DirectiveStart.Anchor -> {
+        val parentList = directives()
+          .filter { it.id == s.parentId }
+          .collect(totalBounds())
+        if (parentList.size != 1) throw ParentSearchException(s.parentId, parentList.size)
+        parentList.first()
+      }
+      is DirectiveStart.Absolute -> null
+    }
+    val resolved = directive.resolve(id, parent)
+    uncommittedChanges.add(Edit.Create(resolved))
+
+    adapter.validate(resolved)
+
+    adapter.create(resolved)
+
+    for (simResults in upToDateSimResultsSet) {
+      simResults.get()?.setStale(true)
+    }
+    // create a new list instead of `.clear` because commit objects have the same reference
+    upToDateSimResultsSet = mutableSetOf()
+
+    return id
+  }
+
+  override fun delete(directive: Directive<AnyDirective>, strategy: DeletedAnchorStrategy) {
```

Comment:
Nonblocking: could you add a couple of inline comments to this function? The outer `else` branch isn't doing much, but that's hard to see without dissecting the Kotlin-isms.

Also L188 is an extra newline.